src/main/java/me/eigenraven/lwjgl3ify/textures/NativeBackedImage.java

```diff
@@ -0,0 +1,147 @@
+        this.height = height;
+        this.pointer = pointer;
+
+        // 4 channels: RGBA
+        this.sizeBytes = width * height * 4;
+    }
+
+    public int getWidth() {
+        return width;
+    }
+
+    public int getHeight() {
+        return height;
+    }
+
+    @Override
+    public int[] getRGB(int startX, int startY, int w, int h, int[] rgbArray, int offset, int scansize) {
+
+        // Inverse itr for cache coherency
+        for (int z = startY; z < h; z++) {
+            for (int x = startX; x < w; x++) {
+                int color = MemoryUtil.memGetInt(this.pointer + ((x + z * width) * 4));
+                // ABGR -> ARGB
+                int a = (color >> 24) & 0xFF;
+                int b = (color >> 16) & 0xFF;
+                int g = (color >> 8) & 0xFF;
+                int r = (color >> 0) & 0xFF;
+
+                int finalColor = (a << 24) | (r << 16) | (g << 8) | b;
+
+                rgbArray[x + (z * width)] = finalColor;
+            }
+        }
+
+        return rgbArray;
+    }
+
+    @Override
+    public int getRGB(int x, int z) {
+        checkBounds(x, z);
+
+        return MemoryUtil.memGetInt(this.pointer + ((x + z * width) * 4));
+    }
+
+    @Override
+    public void setRGB(int x, int z, int rgb) {
+        checkBounds(x, z);
+
+        MemoryUtil.memPutInt(this.pointer + ((x + z * width) * 4), rgb);
+    }
+
+    @Override
+    public BufferedImage getSubimage(int x, int y, int w, int h) {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    @Override
+    public void close() throws Exception {
+        STBImage.nstbi_image_free(this.pointer);
+    }
+
+    private void checkBounds(int x, int z) {
+        if (x < 0 || x >= this.width || z < 0 || z >= this.height) {
+            throw new IllegalStateException("Out of bounds: " + x + ", " + z + " (width: " + this.width + ", height: " + this.height + ")");
+        }
+    }
+
+    // Parsing
+
+    public static NativeBackedImage make(InputStream stream) throws IOException {
+        ByteBuffer imgBuf = null;
+
+        try {
+            imgBuf = readResource(stream);
+            imgBuf.rewind();
+
+            try (MemoryStack memoryStack = MemoryStack.stackPush()) {
+                IntBuffer width = memoryStack.mallocInt(1);
+                IntBuffer height = memoryStack.mallocInt(1);
+                IntBuffer channels = memoryStack.mallocInt(1);
+
+                // 4 channels: RGBA
+                ByteBuffer buf = STBImage.stbi_load_from_memory(imgBuf, width, height, channels, 4);
+                if (buf == null) {
+                    throw new IOException("Could not load image: " + STBImage.stbi_failure_reason());
+                }
+
+                return new NativeBackedImage(
+                    width.get(0), height.get(0), MemoryUtil.memAddress(buf)
+                );
+            }
+
+        } finally {
+            // free
+            MemoryUtil.memFree(imgBuf);
```

Comment:
probably need to check if imgBuf is null, incase an IO error happens while reading resource.